### PR TITLE
expand after expr_idxs in Linearizer.global_load

### PR DESCRIPTION
**Pull request by chaosagent**
This changes the behavior of float4 upcast to either succeed for the whole global load, or fail for the whole global load. There is no difference from previous behavior, at least in CI (see CI for [36396a5](https://github.com/tinygrad/tinygrad/pull/1818/commits/36396a53d2651ae028a9643e08aafd48f313bf89)); no comparison asserts trip)

this + substitute only upcasted + flat_components + cache View.expr_idxs:
```
(venv) chaos@tiny3:~/tinygrad$ STEPS=5 OPT=2 WINO=1 GPU=1 JIT=1 python examples/hlb_cifar10.py
/home/chaos/tinygrad/venv/lib/python3.11/site-packages/pyopencl/cache.py:495: CompilerWarning: Non-empty compiler output encountered. Set the environment variable PYOPENCL_COMPILER_OUTPUT=1 to see more.
  _create_built_program_from_source_cached(
  0 13410.33 ms run, 13410.09 ms python,    0.23 ms CL, 1187.23 loss, 0.002563 LR, 0.84 GB used,     35.40 GFLOPS
  1 4891.89 ms run, 4891.34 ms python,    0.55 ms CL, 1187.59 loss, 0.002577 LR, 4.39 GB used,     95.73 GFLOPS
  2   50.13 ms run,    2.85 ms python,   47.28 ms CL, 2760.24 loss, 0.001741 LR, 4.39 GB used,   9342.29 GFLOPS
  3   46.00 ms run,    2.81 ms python,   43.19 ms CL, 2844.66 loss, 0.000906 LR, 4.39 GB used,  10181.02 GFLOPS
  4   45.74 ms run,    2.76 ms python,   42.98 ms CL, 2117.96 loss, 0.000070 LR, 4.39 GB used,  10239.04 GFLOPS
```

Just this set of changes:
```
(venv) chaos@tiny3:~/tinygrad$ git log | head -n1 && STEPS=10 DEBUG=0 TC=0 OPT=2 WINO=1 GPU=1 JIT=1 python examples/hlb_cifar10.py
commit 25b23bca153f50e343de390a12355ffa79b7f948
/home/chaos/tinygrad/venv/lib/python3.11/site-packages/pyopencl/cache.py:495: CompilerWarning: Non-empty compiler output encountered. Set the environment variable PYOPENCL_COMPILER_OUTPUT=1 to see more.
  _create_built_program_from_source_cached(
  0 14123.57 ms run, 14123.34 ms python,    0.23 ms CL, 1187.23 loss, 0.001281 LR, 0.84 GB used,     33.61 GFLOPS
  1 5468.59 ms run, 5468.02 ms python,    0.57 ms CL, 1187.59 loss, 0.002563 LR, 4.39 GB used,     85.64 GFLOPS
  2   50.60 ms run,    2.78 ms python,   47.82 ms CL, 1903.44 loss, 0.002994 LR, 4.39 GB used,   9255.78 GFLOPS
  3   45.82 ms run,    2.79 ms python,   43.03 ms CL, 2734.18 loss, 0.002577 LR, 4.39 GB used,  10221.41 GFLOPS
  4   45.90 ms run,    5.03 ms python,   40.87 ms CL, 2271.15 loss, 0.002159 LR, 4.39 GB used,  10202.22 GFLOPS
```

Baseline:

```
(venv) chaos@tiny3:~/tinygrad$ STEPS=5 OPT=2 WINO=1 GPU=1 JIT=1 python examples/hlb_cifar10.py
/home/chaos/tinygrad/venv/lib/python3.11/site-packages/pyopencl/cache.py:495: CompilerWarning: Non-empty compiler output encountered. Set the environment variable PYOPENCL_COMPILER_OUTPUT=1 to see more.
  _create_built_program_from_source_cached(
  0 20456.52 ms run, 20456.29 ms python,    0.23 ms CL, 1187.23 loss, 0.002563 LR, 0.84 GB used,     23.21 GFLOPS
  1 6547.99 ms run, 6547.42 ms python,    0.56 ms CL, 1187.59 loss, 0.002577 LR, 4.39 GB used,     71.52 GFLOPS
  2   50.41 ms run,    2.80 ms python,   47.60 ms CL, 2760.24 loss, 0.001741 LR, 4.39 GB used,   9291.11 GFLOPS
  3   45.71 ms run,    2.86 ms python,   42.85 ms CL, 2844.66 loss, 0.000906 LR, 4.39 GB used,  10245.67 GFLOPS
  4   45.36 ms run,    3.86 ms python,   41.50 ms CL, 2117.96 loss, 0.000070 LR, 4.39 GB used,  10325.19 GFLOPS
```

credit to @chenyuxyz for only substituting upcasted axes :D

### CODE REVIEW

**Comment by tinyb0t** [comment]
Changes made in `tinygrad/`:
```
------------------------------------------------------------
files                             insertions       deletions
------------------------------------------------------------
tinygrad/codegen/linearizer.py            17              14
tinygrad/shape/shapetracker.py             2               0
tinygrad/shape/symbolic.py                 8              10
------------------------------------------------------------
total                                     27              24
------------------------------------------------------------
lines added in the tinygrad folder: 3
```

File: tinygrad/shape/symbolic.py [hunk 1]
```diff
@@ -21,7 +21,11 @@ def render(self, ops=None, ctx=None) -> str:
     return ops[type(self)](self, ops, ctx)
   def vars(self): return []
   # expand a Node into List[Node] that enumerates the underlying Variables from min to max
-  def expand(self) -> List[Node]: raise RuntimeError(self.__class__.__name__)
+  @functools.lru_cache(maxsize=None)
+  def expand(self, ctx:Tuple[Variable, ...]=None) -> List[Node]:
+    if ctx is None: ctx = tuple(v for v in set(self.vars()) if v.expr is None)
+    replacements = itertools.product(*[[NumNode(x) for x in range(v.min, v.max + 1)] for v in ctx[::-1]])
```
Comment (roelofvandijk) [inline tinygrad/shape/symbolic.py:27]: Import product from itertools, saves the lookup

File: tinygrad/shape/symbolic.py [hunk 2]
```diff
@@ -21,7 +21,11 @@ def render(self, ops=None, ctx=None) -> str:
     return ops[type(self)](self, ops, ctx)
   def vars(self): return []
   # expand a Node into List[Node] that enumerates the underlying Variables from min to max
-  def expand(self) -> List[Node]: raise RuntimeError(self.__class__.__name__)
+  @functools.lru_cache(maxsize=None)
+  def expand(self, ctx:Tuple[Variable, ...]=None) -> List[Node]:
+    if ctx is None: ctx = tuple(v for v in set(self.vars()) if v.expr is None)
```
Comment (roelofvandijk) [inline tinygrad/shape/symbolic.py:26]: list comprehension into tuples seems faster than comprehension into tuple.

File: tinygrad/shape/symbolic.py [hunk 3]
```diff
@@ -21,7 +21,11 @@ def render(self, ops=None, ctx=None) -> str:
...
+    replacements = itertools.product(*[[NumNode(x) for x in range(v.min, v.max + 1)] for v in ctx[::-1]])
+    return [self.substitute(dict(zip(ctx[::-1], rep))) for rep in replacements]
```
Comment (roelofvandijk) [inline tinygrad/shape/symbolic.py:28]: Can replacements be empty?

Comment (chaosagent) [reply tinygrad/shape/symbolic.py:28]: Nice catch, looks like python is one step ahead though XD
```
>>> list(itertools.product())
[()]
```

Comment (roelofvandijk) [reply tinygrad/shape/symbolic.py:28]: Shouldn't this be an empty list? Does adding a fast path after the None check help?
If not ctx: return []

File: tinygrad/shape/symbolic.py (hunk 1 above)
Comment (roelofvandijk) [inline tinygrad/shape/symbolic.py:27]: Can anyone of them be 0?

File: tinygrad/codegen/linearizer.py [hunk 4]
```diff
@@ -86,34 +86,37 @@ def const(self, b:Union[int,float], dtype=dtypes.int32) -> UOp: return self.uop(
...
     upcast_dim = self.get_upcast_dim(i)
 
-    amt = 1
+    amt, dim = 1, None
     if len(upcast_dim) == 1 and len(expanded_nodes[upcast_dim[0]]) in [4,2]:
       dim, amt = upcast_dim[0], len(expanded_nodes[upcast_dim[0]])
 
     # calculate expr_idxs using placeholder variables
-    fake_idxs = [idx if isinstance(idx, NumNode) else Variable(f"_uidx{i}", idx.min, idx.max) for i, idx in enumerate(idxs)]
+    fake_idxs = [(Variable(f"_uidx{j}", idx.min, idx.max) // (1 if j != dim else amt)) if None in [v.expr for v in idx.vars()] else idx for j, idx in enumerate(idxs)]
```
Comment (gijskoning) [inline tinygrad/codegen/linearizer.py:105]: Avoids iterating full list if None is found
```suggestion
    fake_idxs = [(Variable(f"_uidx{j}", idx.min, idx.max) // (1 if j != dim else amt)) if any(v.expr is None for v in idx.vars()) else idx for j, idx in enumerate(idxs)]
```

Comment (gijskoning) [reply tinygrad/codegen/linearizer.py:105]: Happens more below. Not sure how long idx.vars() is. Memorizing might be nice?

**Comment by chenyuxyz** [comment]
the compile time is faster because it only expands once per `global_load`, but the generated kernel is slower because for some `_idx`, the expand can be further simplified. We lost that info if we only expand once

**Comment by chaosagent** [comment]
> the compile time is faster because it only expands once per `global_load`, but the generated kernel is slower because for some `_idx`, the expand can be further simplified. We lost that info if we only expand once

Do you have an example? In principle this should just be the substitute version but with a cache.

**Comment by chenyuxyz** [comment]
hmm I think you are right for the logic. Somehow I applied the change and got a worse kernel. Let me investigate

**Comment by chaosagent** [comment]
The logic for float4 fallback changed: now global_load only uses float4 if all loads from one buf/st are aligned and contiguous, before it would use float4 for some but not for others. I am curious to see a real example though.

File: tinygrad/shape/symbolic.py (hunk 1 above)
Comment (chaosagent) [reply tinygrad/shape/symbolic.py:27]: in linearizer, when global_load is called, all indices have dimension at least 2.

File: tinygrad/shape/symbolic.py [hunk 5]
```diff
@@ -21,7 +21,18 @@ def render(self, ops=None, ctx=None) -> str:
     return ops[type(self)](self, ops, ctx)
   def vars(self): return []
   # expand a Node into List[Node] that enumerates the underlying Variables from min to max
-  def expand(self) -> List[Node]: raise RuntimeError(self.__class__.__name__)
+  # expand increments earlier variables faster than later variables (as specified in the argument)
+  @functools.lru_cache(maxsize=None)  # pylint: disable=method-cache-max-size-none
```
Comment (chaosagent) [inline tinygrad/shape/symbolic.py:25]: should this stuff be moved out of here? (perhaps back to Linearizer) expand is now just a substitute call in a loop, and thus stuff is cluttering up symbolic.py

File: tinygrad/codegen/linearizer.py [hunk 6]
```diff
@@ -85,35 +84,31 @@ def const(self, b:Union[int,float], dtype=dtypes.int32) -> UOp: return self.uop(
...
 
-    amt = 1
-    if len(upcast_dim) == 1 and len(expanded_nodes[upcast_dim[0]]) in [4,2]:
-      dim, amt = upcast_dim[0], len(expanded_nodes[upcast_dim[0]])
+    amt, dim = 1, None
+    upcast_dim = self.get_upcast_dim(i)
+    if len(upcast_dim) == 1 and len(float4_expand := idxs[upcast_dim[0]].expand()) in [4,2]:
+      dim, amt = upcast_dim[0], len(float4_expand)
+
+    expand_vars = tuple([rename_var(idx.expand_idx(), f"_uidx{j}") for j, idx in enumerate(idxs)])
```
Comment (chaosagent) [inline tinygrad/codegen/linearizer.py:97]: if we only expand only on vars that need expanding, we gain 2.5%, but makes these list comprehensions a little messy

File: tinygrad/shape/symbolic.py (hunk 3 above)
Comment (chaosagent) [reply tinygrad/shape/symbolic.py]: replacements is never empty, since expanding a symbolic expression should at least yield 1 symbolic expression

**Comment by chaosagent** [comment]
I think this change is at a good point. There can be some more simplifying refactors, but perfect is the enemy of the good

**Comment by geohot** [comment]
Are these separate things that can be split into individual PRs?

**Comment by geohot** [comment]
I think this needs a lot more testing

**Comment by chenyuxyz** [comment]
@chaosagent are you finishing this or do you need help finishing it?

**Comment by chaosagent** [comment]
I will finish it, sorry been distracted by some other stuff recently

**Comment by chaosagent** [comment]
I added some tests, seems like the behavior here just depends on unit_stride_axes

**Comment by geohot** [comment]
Neat! Great use of schedule.

You don't actually have to realize the tensors in those tests, just take the [-1] from schedule instead and remove the realize

**Comment by geohot** [comment]
Big speedups for me on `test_winograd.py`

Old:
```
kafka@tubby:~/tinygrad$ python3 test/test_winograd.py 
test_speed (__main__.TestWinograd) ... (3, 3, 4, 4)
running conv: 242.19 ms
scheduling: 8.90 ms
linearize 1 with  911 ops: 148.02 ms
linearize 4 with 1723 ops: 1190.92 ms
linearize 6 with    4 ops: 1.55 ms
linearize 7 with  955 ops: 154.13 ms
linearize 8 with    2 ops: 0.42 ms
```

New:
```
kafka@tubby:~/tinygrad$ python3 test/test_winograd.py 
test_speed (__main__.TestWinograd) ... (3, 3, 4, 4)
running conv: 228.73 ms
scheduling: 8.53 ms
linearize 1 with  911 ops: 106.80 ms
linearize 4 with 1723 ops: 640.88 ms
linearize 6 with    4 ops: 1.49 ms
linearize 7 with  955 ops: 161.23 ms
linearize 8 with    2 ops: 0.45 ms
```

**Pull request merged**